test/subscription_map.h

```diff
@@ -0,0 +1,241 @@
+//
+// Created by wkl04 on 10-12-2019.
+//
+
+#ifndef MQTTSUBSCRIPTION_SUBSCRIPTION_MAP_H
+#define MQTTSUBSCRIPTION_SUBSCRIPTION_MAP_H
+
+#include <mqtt/string_view.hpp>
+
+#include <unordered_map>
+#include "path_tokenizer.h"
+
+template<typename Value>
+class subscription_map_base
+{
+    typedef size_t node_id;
+    typedef std::pair< node_id, std::string> path_entry_key;
+
+    enum { root_node_id = 0 };
+
+    struct path_entry
+    {
+        node_id id;
+
+        bool has_hash_child : 1;
+        bool has_plus_child : 1;
+        uint32_t count : 30;
+
+        Value value;
+
+        path_entry(node_id _id)
+                : id(_id), has_hash_child(false), has_plus_child(false), count(1)
+        { }
+    };
+
+    typedef std::unordered_map< path_entry_key, path_entry, boost::hash< path_entry_key > > map_type;
+    typedef typename map_type::iterator map_type_iterator;
+    typedef typename map_type::const_iterator map_type_const_iterator;
+
+    map_type map;
+    map_type_iterator root;
+    node_id next_node_id;
+
+protected:
+    map_type_iterator end() { return map.end(); }
+
+    std::vector< std::pair<map_type_iterator, map_type_iterator> > find_subscription(MQTT_NS::string_view const &topic)
+    {
+        auto tokens = mqtt_path_tokenizer(topic);
+        auto parent = root;
+
+        std::vector< std::pair<map_type_iterator, map_type_iterator> > path;
+
+        for (auto const  &t : tokens) {
+            auto entry = map.find(path_entry_key(parent->second.id, t));
+
+            if(entry == map.end())
```

Comment:
Don't omit curly braces

Reply:
So preferred style is to always use curly braces ? Also for one line control statements ?

Reply:
Omitting curley braces is a security problem.

See: OpenSSL's Heartbleed vulnerability.

They should never be omitted, regardless of the coding style of the code being worked on.